File: lib/video_transformer.dart
import 'aft_transformer_encoder.dart';
import 'gpu_tensor.dart';
import 'nn.dart';

class VideoTransformer extends Module {
  // Inherit from Module for the override
  final int frameEmbedDim;
  final int embedSize;
  final int maxVideoSequenceLength;
  final int numClasses;

  final Layer? frameProjection;
  final Tensor
  posEmbeddings; // Renamed from positionEmbeddings to match parameters()
  final TransformerEncoder transformerEncoder;
  final Layer mlpHead;

  VideoTransformer({
    required this.frameEmbedDim,
    required this.embedSize,
    required this.maxVideoSequenceLength,
    required this.numClasses,
    int numLayers = 2,
    int numHeads = 4,
  }) : frameProjection = (frameEmbedDim != embedSize)
           ? Layer(frameEmbedDim, embedSize)
           : null,
       posEmbeddings = Tensor.random([maxVideoSequenceLength, embedSize]),
       transformerEncoder = TransformerEncoder(
         vocabSize: 0, // Added to match standard signature
         embedSize: embedSize,
         blockSize: maxVideoSequenceLength,
         numLayers: numLayers,
         numHeads: numHeads,
       ),
       mlpHead = Layer(embedSize, numClasses);

  /// Forward pass on the GPU
  Tensor forward(Tensor videoEmbeddings, List<Tensor> tracker) {
    final int numFrames = videoEmbeddings.shape[0];

    if (numFrames > maxVideoSequenceLength) {
      throw ArgumentError("Video too long for maxVideoSequenceLength");
    }

    // 1. Projection
    Tensor x = frameProjection != null
        ? frameProjection!.forward(videoEmbeddings, tracker)
        : videoEmbeddings;

    // 2. Slice and Add Positions
    final pos = posEmbeddings.slice(0, numFrames);
    // Note: slice is a view, usually doesn't need tracking unless you want to auto-dispose

    final xWithPos = x + pos;
    tracker.add(xWithPos);

    // 3. Transformer Encoder
    final encoded = transformerEncoder.forwardEmbeddings(xWithPos, tracker);

    // 4. Global Average Pooling
    final pooled = encoded.mean();
    tracker.add(pooled);

    // 5. MLP Head
    final logits = mlpHead.forward(pooled, tracker);

    return logits;
  }

  @override
  List<Tensor> parameters() {
    return [
      if (frameProjection != null) ...frameProjection!.parameters(),
      posEmbeddings, // Now this matches the field name above!
      ...transformerEncoder.parameters(),
      ...mlpHead.parameters(),
    ];
  }
}

File: python/dart_cuda/core/transformers/modalities/video_transformer.py
"""Video transformer — mirrors `modalities/video_transformer.dart`."""

from __future__ import annotations

from typing import List, Optional

from ...layers.nn import Layer, Module
from ...tensor.gpu_tensor import Tensor
from ..aft.transformer_encoder import TransformerEncoder


class VideoTransformer(Module):
    def __init__(
        self,
        *,
        frameEmbedDim: int,
        embedSize: int,
        maxVideoSequenceLength: int,
        numClasses: int,
        numLayers: int = 2,
        numHeads: int = 4,
    ):
        self.frame_embed_dim = frameEmbedDim
        self.embed_size = embedSize
        self.max_video_sequence_length = maxVideoSequenceLength
        self.num_classes = numClasses

        self.frame_projection: Optional[Layer]
        if frameEmbedDim != embedSize:
            self.frame_projection = Layer(frameEmbedDim, embedSize, useGelu=False)
        else:
            self.frame_projection = None
        self.pos_embeddings = Tensor.random([maxVideoSequenceLength, embedSize])
        self.transformer_encoder = TransformerEncoder(
            vocabSize=0,
            embedSize=embedSize,
            blockSize=maxVideoSequenceLength,
            numLayers=numLayers,
            numHeads=numHeads,
        )
        self.mlp_head = Layer(embedSize, numClasses, useGelu=False)

    @property
    def embedSize(self) -> int:
        return self.embed_size

    @property
    def frameProjection(self) -> Optional[Layer]:
        return self.frame_projection

    @property
    def posEmbeddings(self) -> Tensor:
        return self.pos_embeddings

    @property
    def transformerEncoder(self) -> TransformerEncoder:
        return self.transformer_encoder

    def forward(self, video_embeddings: Tensor, tracker: List[Tensor]) -> Tensor:
        num_frames = video_embeddings.shape[0]
        if num_frames > self.max_video_sequence_length:
            raise ValueError("Video too long for maxVideoSequenceLength")
        x = (
            self.frame_projection.forward(video_embeddings, tracker)
            if self.frame_projection is not None
            else video_embeddings
        )
        pos = self.pos_embeddings.slice(0, num_frames)
        x_with_pos = x + pos
        tracker.append(x_with_pos)
        encoded = self.transformer_encoder.forward_embeddings(x_with_pos, tracker)
        pooled = encoded.mean()
        tracker.append(pooled)
        return self.mlp_head.forward(pooled, tracker)

    def parameters(self) -> List[Tensor]:
        ps: List[Tensor] = []
        if self.frame_projection is not None:
            ps.extend(self.frame_projection.parameters())
        ps.append(self.pos_embeddings)
        ps.extend(self.transformer_encoder.parameters())
        ps.extend(self.mlp_head.parameters())
        return ps
